sync/src/orphan_block_pool.rs

```diff
@@ -30,10 +32,20 @@ impl OrphanBlockPool {
             .entry(parent_hash.clone())
             .or_insert_with(HashMap::default)
             .insert(hash.clone(), block);
-        self.parents.insert(hash, parent_hash);
+        // Out-of-order insertion needs to be deduplicated
+        self.leaders.remove(&hash);
+        self.parents.insert(hash, parent_hash.clone());
+        if !self.parents.contains_key(&parent_hash) {
+            self.leaders.insert(parent_hash);
+        }
     }
 
     pub fn remove_blocks_by_parent(&self, hash: &packed::Byte32) -> Vec<core::BlockView> {
```

Comment:
Since the hash intends to be a parent hash, it's better to make it explicitly

```suggestion
    pub fn remove_blocks_by_parent(&self, parent_hash: &ParentHash) -> Vec<core::BlockView> {
```
```